### PR TITLE
Fix next major comment

### PR DESCRIPTION
**Pull request by VincentLanglet**

### CODE REVIEW

**Comment by VincentLanglet** [comment]
@wbloszyk In AddDependencyCallCompilerPass there is the following code
```
$templateRegistryId = sprintf('%s.template_registry', $serviceId);
        $templateRegistryDefinition = $container
            ->register($templateRegistryId, TemplateRegistry::class)
            ->addTag('sonata.admin.template_registry')
            ->setPublic(true); // Temporary fix until we can support service locators

        if ($container->getParameter('sonata.admin.configuration.templates') !== $definedTemplates) {
            $templateRegistryDefinition->addArgument($definedTemplates);
        } else {
            $templateRegistryDefinition->addArgument('%sonata.admin.configuration.templates%');
        }

        $definition->addMethodCall('setTemplateRegistry', [new Reference($templateRegistryId)]);
```

I followed your NEXT_MAJOR comment and now It seems like we pass a templateRegistry instead of an mutable one. What should be changed to fix this ?

File: docs/reference/templates.rst [hunk 1]
```diff
@@ -194,11 +194,7 @@ can specify the templates to use in the ``Admin`` service definition:
 .. note::
 
     A ``setTemplates(array $templates)`` (notice the plural) method also
-    exists, that allows you to set multiple templates at once. Notice that,
-    if used outside of the service definition context,
-    ``setTemplates(array $templates)`` will replace the whole template list
-    for that ``Admin`` class, meaning you have to explicitly pass the full
-    template list in the ``$templates`` argument.
+    exists, that allows you to set multiple templates at once.
```
Comment (wbloszyk) [inline docs/reference/templates.rst:197]: We are changing override by merge. IMO will be nice to add note about it for people who know about previus functionality.

File: tests/DependencyInjection/SonataAdminExtensionTest.php [hunk 2]
```diff
@@ -94,10 +93,7 @@ public function testHasCoreServicesAlias(): void
             SessionFilterPersister::class
         );
         $this->assertContainerBuilderHasService(TemplateRegistry::class);
-        $this->assertContainerBuilderHasService(
-            MutableTemplateRegistryInterface::class,
-            TemplateRegistry::class
-        );
+        $this->assertContainerBuilderHasService(TemplateRegistry::class);
```
Comment (wbloszyk) [inline tests/DependencyInjection/SonataAdminExtensionTest.php:96]:
```suggestion
```
This line is duplicated.

File: src/Admin/AbstractAdmin.php [hunk 3]
```diff
@@ -886,12 +866,12 @@ public function getNewInstance(): object
...
-        $this->formOptions['data_class'] = $this->getClass();
-
         $formBuilder = $this->getFormContractor()->getFormBuilder(
             $this->getUniqid(),
-            // NEXT_MAJOR : remove the merge with $this->formOptions
-            array_merge($this->getFormOptions(), $this->formOptions)
+            array_merge(
+                $this->getFormOptions(),
+                ['data_class' => $this->getClass()]
+            )
```
Comment (wbloszyk) [inline src/Admin/AbstractAdmin.php:874]:
```suggestion
            ['data_class' => $this->getClass()] + $this->getFormOptions()
```

**Review by wbloszyk — CHANGES_REQUESTED** [review]

**Comment by VincentLanglet** [comment]
What should I do about the test failure @wbloszyk ?

**Comment by wbloszyk** [comment]
> @wbloszyk In AddDependencyCallCompilerPass there is the following code
> 
> ```
> $templateRegistryId = sprintf('%s.template_registry', $serviceId);
>         $templateRegistryDefinition = $container
>             ->register($templateRegistryId, TemplateRegistry::class)
>             ->addTag('sonata.admin.template_registry')
>             ->setPublic(true); // Temporary fix until we can support service locators
> 
>         if ($container->getParameter('sonata.admin.configuration.templates') !== $definedTemplates) {
>             $templateRegistryDefinition->addArgument($definedTemplates);
>         } else {
>             $templateRegistryDefinition->addArgument('%sonata.admin.configuration.templates%');
>         }
> 
>         $definition->addMethodCall('setTemplateRegistry', [new Reference($templateRegistryId)]);
> ```
> 
> I followed your NEXT_MAJOR comment and now It seems like we pass a templateRegistry instead of an mutable one. What should be changed to fix this ?

```php
 $templateRegistryDefinition = $container
     ->register($templateRegistryId, MutableTemplateRegistry::class)
     ->addTag('sonata.admin.template_registry')
     ->setPublic(true); // Temporary fix until we can support service locato
```

**Comment by wbloszyk** [comment]
@VincentLanglet 
I also think about create `MutableTemplateRegistry` in `AbstractAdmin::__construtor()`. In this case we do not have to create template registry for each admin. Instead we can use `setTemplates()` which will be modify by add next parameter `setTemplates(array $templates, bool $override = true)`. Then in `AddDependencyCallCompilerPass` we will call `setTemplates()` to add missing templates only.

File: docs/reference/templates.rst (hunk 1 above)
Comment (VincentLanglet) [reply docs/reference/templates.rst:197]: I added a note in the upgrade note

File: tests/DependencyInjection/SonataAdminExtensionTest.php (hunk 2 above)
Comment (VincentLanglet) [reply tests/DependencyInjection/SonataAdminExtensionTest.php]: done

File: src/Admin/AbstractAdmin.php (hunk 3 above)
Comment (VincentLanglet) [reply src/Admin/AbstractAdmin.php]: done

**Review by wbloszyk — APPROVED** [review]

**Review by wbloszyk — APPROVED** [review]

**Review by dmaicher — APPROVED** [review]

**Pull request merged**